### ISSUE TITLE
False positive when iterating over a static list of Literals

### ISSUE THREAD

**Issue by sfo** [microsoft/pyright#6397]
**Describe the bug**
Pyright fails to infer the correct type from a static list of `Literal[..]`-typed elements when iterating over it.
The code below shows some examples and indicates the place, at which pylance reports the following error:

```
Argument of type "str" cannot be assigned to parameter "x" of type "Literal['foo', 'bar']" in function "foobar"
  Type "str" cannot be assigned to type "Literal['foo', 'bar']"
    "str" cannot be assigned to type "Literal['foo']"
    "str" cannot be assigned to type "Literal['bar']"
```

I assume, it should be possible for a static type checker to infer the correct type of `str` in this case, which should be something like `Literal[ <list content> ]`.

**Code or Screenshots**
```python
from typing import Literal


def foobar(x: Literal["foo", "bar"]):
    return x.upper()


print(foobar("foo"))  # <-- true negative
print(foobar("bar"))  # <-- true negative
print(foobar("baz"))  # <-- true positive

for str in ["foo", "bar"]:
    print(foobar(str))  # <-- false positive

for str in ["foo", "bar"]:
    str = "foo"
    print(foobar(str))  # <-- true negative
```

**VS Code extension or command-line**
Version: 1.85.0-insider
Commit: e55009850900a64eeb37b2228c79855ceff63914
Date: 2023-11-03T15:12:10.231Z
Electron: 25.9.2
ElectronBuildId: 24603566
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Linux x64 6.5.9-200.fc38.x86_64 snap

Pylance: 2023.11.12

**Comment by erictraut** [comment]
Pyright's behavior is correct here. The type of the expression `["foo", "bar"]` is `list[str]`. Literals are not retained for lists because they are mutable. If you change this to a tuple, literals will be retained.

```python
for str in ("foo", "bar"):
    print(foobar(str))

for str in ("foo", "bar"):
    str = "foo"
    print(foobar(str))
```

**Comment by erictraut** [comment]
For more details about pyright's type inference behavior, refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=literals).
